wiring/inc/spark_wiring_cloud_event.h

```diff
@@ -0,0 +1,477 @@
+     *
+     * @param event Event to copy.
+     */
+    CloudEvent(const CloudEvent& event);
+
+    /**
+     * Move constructor.
+     *
+     * @param event Event to move from.
+     */
+    CloudEvent(CloudEvent&& event);
+
+    /**
+     * Destructor.
+     *
+     * Destroying an event that is being sent to the Cloud does not cancel the sending operation.
+     *
+     * @see `cancel()`.
+     */
+    ~CloudEvent();
+
+    /**
+     * Set the event name.
+     *
+     * @param name Event name.
+     * @return This event instance.
+     */
+    CloudEvent& name(const char* name);
+
+    /**
+     * Get the event name.
+     *
+     * @return Event name.
+     */
+    const char* name() const;
+
+    /**
+     * Set the content type of the event data.
+     *
+     * @param type Content type.
+     * @return This event instance.
+     */
+    CloudEvent& contentType(ContentType type);
+
+    /**
+     * Get the content type of the event data.
+     *
+     * @return Content type.
+     */
+    ContentType contentType() const;
+
+    /**
+     * Set the event data.
+     *
+     * @param data Event data.
+     * @return This event instance.
+     */
+    CloudEvent& data(const char* data) {
+        return this->data(data, std::strlen(data));
+    }
+
+    /**
+     * Set the event data.
+     *
+     * @param data Event data.
+     * @param data Data size.
+     * @return This event instance.
+     */
+    CloudEvent& data(const char* data, size_t size);
+
+    /**
+     * Set the event data.
+     *
+     * @param data Event data.
+     * @param data Data size.
+     * @param type Content type.
+     * @return This event instance.
+     */
+    CloudEvent& data(const char* data, size_t size, ContentType type) {
+        this->data(data, size);
+        contentType(type);
+        return *this;
+    }
+
+    /**
+     * Set the event data.
+     *
+     * @param data Event data.
+     * @return This event instance.
+     */
+    CloudEvent& data(const Buffer& data) {
+        return this->data(data.data(), data.size());
+    }
+
+    /**
+     * Set the event data.
+     *
+     * @param data Event data.
+     * @param type Content type.
+     * @return This event instance.
+     */
+    CloudEvent& data(const Buffer& data, ContentType type) {
+        return this->data(data.data(), data.size(), type);
+    }
+
+    /**
+     * Set the event data.
+     *
+     * @param data Event data.
+     * @return This event instance.
+     */
+    CloudEvent& data(const Variant& data);
+
+    /**
+     * Get the event data.
+     *
+     * This method returns a copy of the event data in a dynamically allocated buffer.
+     *
+     * @return Event data.
+     */
+    Buffer data() const;
+
+    /**
+     * Get the event data.
+     *
+     * @return Event data.
+     */
+    String dataAsString() const;
+
+    /**
+     * Get the event data.
+     *
+     * @return Event data.
+     */
+    Variant dataAsVariant() /* FIXME: const */; // TODO: Rename?
```

Comment:
Is this still relevant since we only support receiving binary payloads from large events?

Reply:
I mostly added it for feature parity with the existing API. The user can still send CBOR data in a binary event and use this method to parse it. Let me know if we don't want to support encoding and decoding variants in `CloudEvent` directly and I'll remove the related methods.

Reply:
Fixed in https://github.com/particle-iot/device-os/pull/2840/commits/6b4d48953f2d8d06923d6e80680db5057faa4562.